### PR TITLE
[video_player] Support http headers while constructing video controller from network datasource.

### PR DESCRIPTION
**Pull request by JarvanMo**
Related #16466.

## Description

Http headers are very important while playing video in a real world.
This PR allow developers adding headers into the video player while playing.

## Related Issues

flutter/flutter#16466.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by JohnGalt1717** [comment]
Please merge this ASAP!

**Comment by TRUSTMEIMJEDI** [comment]
Are there any news about merging it to the master need this functionality for my engineer thesis ASAP ?

File: packages/video_player/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java [hunk 1]
```diff
@@ -87,13 +88,19 @@
 
       DataSource.Factory dataSourceFactory;
       if (isHTTP(uri)) {
-        dataSourceFactory =
+        DefaultHttpDataSourceFactory defaultHttpDataSourceFactory =
```
[the remark this answers is not in the document]
Comment (JarvanMo) [reply packages/video_player/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java]: Sure.

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m [hunk 2]
```diff
@@ -157,8 +159,17 @@ - (void)createVideoOutputAndDisplayLink:(FLTFrameUpdater*)frameUpdater {
...
-  AVPlayerItem* item = [AVPlayerItem playerItemWithURL:url];
+- (instancetype)initWithURL:(NSURL*)url
+               frameUpdater:(FLTFrameUpdater*)frameUpdater
+                    headers:(NSDictionary*)headers {
+  AVPlayerItem* item;
+  if (headers == (id)[NSNull null] || headers == nil) {
+    item = [AVPlayerItem playerItemWithURL:url];
+  } else {
+    AVURLAsset* asset = [AVURLAsset URLAssetWithURL:url
+                                            options:@{@"AVURLAssetHTTPHeaderFieldsKey" : headers}];
```
[the remark this answers is not in the document]
Comment (JarvanMo) [reply packages/video_player/ios/Classes/VideoPlayerPlugin.m]: It's a hack way.  See [add-custom-header-field-in-request-of-avplayer](https://stackoverflow.com/questions/15456130/add-custom-header-field-in-request-of-avplayer) for detail.

**Comment by JohnGalt1717** [comment]
Any chance we can get web compatibility added as well? (And note macos, Linux and windows support in the future too.)

**Comment by JarvanMo** [comment]
> Any chance we can get web compatibility added as well? (And note macos, Linux and windows support in the future too.)

I'm not sure.  I'm not familiar those platforms.

File: packages/video_player/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (JarvanMo) [reply packages/video_player/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java]: I notice that `getDefaultRequestProperties.set` comes from `HttpDataSource`. We can't call `getDefaultRequestProperties.set`  in `DataSource.Factory` directly.

**Comment by otopba** [comment]
@JarvanMo 
Hi! Could you merge original master to this PR ? Thank you

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (sanekyy) [reply packages/video_player/ios/Classes/VideoPlayerPlugin.m]: is it a private API as said [here](https://github.com/flutter/flutter/issues/16466#issuecomment-574075820)?

**Comment by sanekyy** [comment]
Is there usage of iOS private API as said [here](https://github.com/flutter/flutter/issues/16466#issuecomment-574075820) and [here](https://stackoverflow.com/a/23713028)?

If its true and apps may reject by AppStore if we use this and this is only way to set https header then I think that we can't accept this PR.

**Comment by sanekyy** [comment]
Hi @JarvanMo.

Is there usage of iOS private API as said [here](https://github.com/flutter/flutter/issues/16466#issuecomment-574075820) and [here](https://stackoverflow.com/a/23713028)?

If its true and apps may reject by AppStore if we use this and this is only way to set https header then I think that we can't accept this PR.

**Comment by Aljabri-Salman** [comment]
we really need this feature it is very important. thanks

**Comment by JohnGalt1717** [comment]
FYI, this can be done without having to use a hidden api call on iOS in a very similar fashion to how UWP works.

**Comment by dvirgiln** [comment]
Hi, 

Any idea when this functionality will be merged? We really need this for our project.

Thanks a lot.

**Pull request closed without merging**